cmd/kubeadm/app/phases/upgrade/health.go

```diff
@@ -0,0 +1,221 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+
+	"k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/cmd/kubeadm/app/constants"
+)
+
+// healthCheck is a helper struct for easily performing healthchecks against the cluster and printing the output
+type healthCheck struct {
+	description, okMessage, failMessage string
+	// f is invoked with a k8s client passed to it. Should return an optional warning and/or an error
+	f func(clientset.Interface) error
+}
+
+// VerifyClusterHealth makes sure:
+// - the API /healthz endpoint is healthy
+// - all Nodes are Ready
+// - (if self-hosted) that there are DaemonSets with at least one Pod for all control plane components
+// - (if static pod-hosted) that all required Static Pod manifests exist on disk
+func VerifyClusterHealth(client clientset.Interface) bool {
+	fmt.Println("[upgrade] Making sure the cluster is healthy:")
+
+	defaultHealthChecks := []healthCheck{
+		{
+			description: "API Server health",
+			okMessage:   "Healthy",
+			failMessage: "Unhealthy",
+			f:           apiServerHealthy,
+		},
+		{
+			description: "Node health",
+			okMessage:   "All Nodes are healthy",
+			failMessage: "More than one Node unhealthy",
+			f:           nodesHealthy,
+		},
+		// TODO: Add a check for ComponentStatuses here?
+	}
+
+	success := runHealthChecks(client, defaultHealthChecks)
+	if !success {
+		return false
+	}
+
+	// If the control plane is self-hosted, we should run more health checks
+	fmt.Printf("[upgrade/health] Checking if control plane is Static Pod-hosted or Self-Hosted: ")
+	if IsControlPlaneSelfHosted(client) {
+
+		fmt.Println("Self-Hosted.")
+
+		// Run an extra pair of health checks to make sure the self-hosted control plane is healthy
+		success = runHealthChecks(client, []healthCheck{
+			{
+				description: "Control plane DaemonSet health",
+				okMessage:   "All control plane DaemonSets are healthy",
+				failMessage: "More than one control plane DaemonSet unhealthy",
+				f:           controlPlaneHealth,
+			},
+		})
+		if !success {
+			return false
+		}
+	} else {
+		fmt.Println("Static Pod-hosted.")
+		fmt.Println("[upgrade/health] NOTE: kubeadm will upgrade your Static Pod-hosted control plane to a Self-Hosted one when upgrading if --feature-gates=SelfHosting=true is set (which is the default)")
+		fmt.Println("[upgrade/health] If you strictly want to continue using a Static Pod-hosted control plane, set --feature-gates=SelfHosting=true when running 'kubeadm upgrade apply'")
```

Comment:
Totally missed that's not the case. I'm aiming to expose the feature flags in the CLI on both `init` and `upgrade`. Good catch